### PR TITLE
Add hash256 and hash160, update ApplicationEngine

### LINKED ISSUE #288: Add RIPEMD160 , hash160, hash256 method
_Add Hash160 and Hash256 to make some Neo2 contract function compatible in Neo3. Original posted at https://github.com/neo-project/neo/pull/1694#issue-432799843_

1. Add `RIPEMD160` syscall
2. add hash160, hash256 as a sdk function

### PR DESCRIPTION
**Pull request by ShawnYun**
close #288 
Link https://github.com/neo-project/neo/pull/1655

### CODE REVIEW

File: src/Neo.SmartContract.Framework/Services/Neo/Crypto.cs [hunk 1]
```diff
@@ -5,23 +5,34 @@ public static class Crypto
...
 
+        [Syscall("Neo.Crypto.RIPEMD160")]
+        public extern static byte[] RIPEMD160(byte[] value);
+
+        [Syscall("Neo.Crypto.SHA256")]
+        [Syscall("Neo.Crypto.RIPEMD160")]
+        public extern static byte[] HASH160(byte[] value);
+
+        [Syscall("Neo.Crypto.SHA256")]
+        [Syscall("Neo.Crypto.SHA256")]
```
Comment (ProDog) [inline src/Neo.SmartContract.Framework/Services/Neo/Crypto.cs:16]:
```suggestion
        [Syscall("Neo.Crypto.SHA256")]
```

File: src/Neo.SmartContract.Framework/Services/Neo/Crypto.cs [hunk 2]
```diff
@@ -5,23 +5,34 @@ public static class Crypto
         [Syscall("Neo.Crypto.SHA256")]
         public extern static byte[] SHA256(byte[] value);
 
+        [Syscall("Neo.Crypto.RIPEMD160")]
+        public extern static byte[] RIPEMD160(byte[] value);
+
+        [Syscall("Neo.Crypto.SHA256")]
+        [Syscall("Neo.Crypto.RIPEMD160")]
```
Comment (ProDog) [inline src/Neo.SmartContract.Framework/Services/Neo/Crypto.cs:12]:
```suggestion
        [Syscall("Neo.Crypto.RIPEMD160")]
```

Comment (ShawnYun) [reply src/Neo.SmartContract.Framework/Services/Neo/Crypto.cs:12]: HASH160 = SHA256().RIPEMD16()  
HASH256 = SHA256().SHA256() 
right?

Comment (ProDog) [reply src/Neo.SmartContract.Framework/Services/Neo/Crypto.cs:12]: Oh, you are right.

**Comment by ProDog** [comment]
I have tested, it's all OK. 

Contract:
```C#
        public static object Hash160Test(byte[] msg)
        {
            Runtime.Notify(Crypto.RIPEMD160(msg));
            Runtime.Notify(Crypto.HASH160(msg));
            return Crypto.HASH160(msg);
        }

        public static object Hash256Test(byte[] msg)
        {
            Runtime.Notify(Crypto.HASH256(msg));
            Runtime.Notify(Crypto.SHA256(msg));
            return Crypto.HASH256(msg);
        }
```

Invoke:
```
neo> invoke 0xa723a2d956ccae9d73c5ef03c300a08a0b6ed796 hash256Test [{"type":"Byte","value":""}]
Invoking script with: '0c0011c00c0b68617368323536546573740c1496d76e0b8aa000c303efc5739daecc56d9a223a741627d5b52'
VM State: HALT
Gas Consumed: 0.0858123
Result Stack: [{"type":"ByteString","value":"Xfbg4nYTWdMKgnUFjimfzAOBU0VF9Vz0PkGYP11MlFY="}]
Relay tx(no|yes): yes
Signed and relayed transaction with hash=0x4c669fa06bfbcae88b4425daae33b88ee9b6b5613d95e284340b4c08ccf44923
neo> invoke 0xa723a2d956ccae9d73c5ef03c300a08a0b6ed796 hash160Test [{"type":"Byte","value":""}]
Invoking script with: '0c0011c00c0b68617368313630546573740c1496d76e0b8aa000c303efc5739daecc56d9a223a741627d5b52'
VM State: HALT
Gas Consumed: 0.0858123
Result Stack: [{"type":"ByteString","value":"tHKiZtC9icE3BqQTLM\u002Bxb3w7n8s="}]
Relay tx(no|yes): yes
Signed and relayed transaction with hash=0x830d40a1f37c899c2d91cf26a52e5f4b61996491f5899eb3355fb5c72613cfd0
```
Notify:
```
"notifications": [
            {
                "contract": "0xa723a2d956ccae9d73c5ef03c300a08a0b6ed796",
                "state": {
                    "type": "Array",
                    "value": [
                        {
                            "type": "ByteString",
                            "value": "Xfbg4nYTWdMKgnUFjimfzAOBU0VF9Vz0PkGYP11MlFY="
                        }
                    ]
                }
            },
            {
                "contract": "0xa723a2d956ccae9d73c5ef03c300a08a0b6ed796",
                "state": {
                    "type": "Array",
                    "value": [
                        {
                            "type": "ByteString",
                            "value": "47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU="
                        }
                    ]
                }
            }
        ]

```

```
"notifications": [
            {
                "contract": "0xa723a2d956ccae9d73c5ef03c300a08a0b6ed796",
                "state": {
                    "type": "Array",
                    "value": [
                        {
                            "type": "ByteString",
                            "value": "nBGFpcXp/FRhKAiXfuj1SLIljTE="
                        }
                    ]
                }
            },
            {
                "contract": "0xa723a2d956ccae9d73c5ef03c300a08a0b6ed796",
                "state": {
                    "type": "Array",
                    "value": [
                        {
                            "type": "ByteString",
                            "value": "tHKiZtC9icE3BqQTLM+xb3w7n8s="
                        }
                    ]
                }
            }
        ]
```

File: src/Neo.SmartContract.Framework/Services/Neo/Crypto.cs [hunk 3]
```diff
@@ -5,23 +5,34 @@ public static class Crypto
...
+        public extern static byte[] HASH160(byte[] value);
+
+        [Syscall("Neo.Crypto.SHA256")]
+        [Syscall("Neo.Crypto.SHA256")]
+        public extern static byte[] HASH256(byte[] value);
```
Comment (Tommo-L) [inline src/Neo.SmartContract.Framework/Services/Neo/Crypto.cs:17]:
```suggestion
        public extern static byte[] Hash256(byte[] value);
```

File: src/Neo.SmartContract.Framework/Services/Neo/Crypto.cs [hunk 4]
```diff
@@ -5,23 +5,34 @@ public static class Crypto
         [Syscall("Neo.Crypto.SHA256")]
         public extern static byte[] SHA256(byte[] value);
 
+        [Syscall("Neo.Crypto.RIPEMD160")]
+        public extern static byte[] RIPEMD160(byte[] value);
+
+        [Syscall("Neo.Crypto.SHA256")]
+        [Syscall("Neo.Crypto.RIPEMD160")]
+        public extern static byte[] HASH160(byte[] value);
```
Comment (Tommo-L) [inline src/Neo.SmartContract.Framework/Services/Neo/Crypto.cs:13]:
```suggestion
        public extern static byte[] Hash160(byte[] value);
```

**Pull request merged**